### PR TITLE
documenting autosummarizer, for v7

### PR DESCRIPTION
**Pull request by shellygr**
feature in development currently

(alex: permitting myself to add these links here...)
https://certora-certora-prover-documentation--222.com.readthedocs.build/en/222/docs/prover/cli/options.html#auto-nondet-difficult-internal-funcs
https://certora-certora-prover-documentation--222.com.readthedocs.build/en/222/docs/prover/cli/options.html#auto-nondet-difficult-internal-funcs

### CODE REVIEW

File: docs/user-guide/timeouts/timeouts-main.md [hunk 1]
```diff
@@ -213,6 +213,48 @@ There is also some helpful information in the section on
...
 
+(detect-candidates-for-summarization)=
+## Detect candidates for summarization
+
+In a large codebase it can be hard to find all the functions that may be difficult for the Prover.
+A traditional approach would be to run a simple parametric rule to explore all functions in the 
+relevant contracts, and studying resulting potential timeouts. 
+However such an approach prolongs the feedback loop of working with the Prover.
+
+As an alternative approach, the Prover supports an {ref}`overapproximating <glossary>` _auto-summarization_ mode.
```
Comment (alexandernutz) [inline docs/user-guide/timeouts/timeouts-main.md:224]: I always use "{term}`overapproximating`", tioli of course

File: docs/user-guide/timeouts/timeouts-main.md [hunk 2]
```diff
@@ -213,6 +213,48 @@ There is also some helpful information in the section on
...
+It is based on the idea that internal view or pure functions (in Solidity) that are analyzed
```
Comment (alexandernutz) [inline docs/user-guide/timeouts/timeouts-main.md:225]: to me code-formatting `view` and `pure` seems better (gut-feeling..)

File: docs/user-guide/timeouts/timeouts-main.md [hunk 3]
```diff
@@ -213,6 +213,48 @@ There is also some helpful information in the section on
...
+and found to be heuristically difficult for the Prover can be automatically summarized as `NONDET`, 
+resulting in two positive outcomes:
+1. The run is faster since complex code is summarized early in the Prover's pipeline
+2. The Prover emits the list of _new_ summaries it auto-generated, so that the user can then adapt the list
```
Comment (alexandernutz) [inline docs/user-guide/timeouts/timeouts-main.md:229]: the "new" seems more confusing than helping to me -- if it's important (like you want to emphasize that existing summaries have precedence), just the word is not enough, but I feel like you don't need to mention that here

Comment (alexandernutz) [reply docs/user-guide/timeouts/timeouts-main.md:229]: also, 1. lacks a `.` in the end

File: docs/user-guide/timeouts/timeouts-main.md [hunk 4]
```diff
@@ -213,6 +213,48 @@ There is also some helpful information in the section on
...
+and make the summaries more precise, or remove them altogether if the user wishes so.
+
+The Prover will not auto-summarize methods that were already summarized by the user.
+
+To enable this mode, add {ref}`--auto_nondet_difficult_internal_funcs` to the `certoraRun` command.
+The minimal difficulty threshold used for the auto-summarization
+can be adjusted using {ref}`--auto_nondet_minimal_difficulty`.
```
Comment (alexandernutz) [inline docs/user-guide/timeouts/timeouts-main.md:236]: (I guess one day we will publish the difficulty heuristic, but I also think it's not quite ready yet ...)

**Review by alexandernutz — COMMENTED** [review]
very nice doc for a very nice feature :-) 
some language nitpicking

File: docs/prover/cli/options.md [hunk 5]
```diff
@@ -513,6 +513,44 @@ certoraRun Bank.sol --verify Bank:Bank.spec --summary_recursion_limit 3
...
+### `--auto_nondet_minimal_difficulty`
+
+**What does it do?**
+Sets the minimal difficulty threshold for the auto-summarization mode enabled by {ref}`--auto_nondet_difficult_internal_funcs`.
+
+**When to use it**
+If the results of an initial run with {ref}`--auto_nondet_difficult_internal_funcs` were unsatisfactory,
+one can adjust the default threshold to apply the auto-summarization to potentially more or fewer internal functions.
+
+The notification in the rule report that contains the applied summaries will present the current threshold used by the Prover.
```
[the remark this answers is not in the document]
Comment (shellygr) [reply docs/prover/cli/options.md:546]: I'm indeed afraid to write it here and then it'll change. The Prover displays the value in any case

File: docs/prover/cli/options.md [hunk 6]
```diff
@@ -513,6 +513,44 @@ certoraRun Bank.sol --verify Bank:Bank.spec --summary_recursion_limit 3
...
+**What does it do?**
+Auto-summarizes view or pure internal functions that return a value-type and are
+currently not summarized, and that are found to be heuristically difficult
+for the Prover.
+
+For more information, see {ref}`detect-candidates-for-summarization`.
+
+**When to use it**
+It is recommended when beginning to work on a large codebase in order
+to get relatively faster feedback from the Prover on potentially difficult functions.
```
[the remark this answers is not in the document]
Comment (shellygr) [reply docs/prover/cli/options.md:528]: thanks

File: docs/user-guide/timeouts/timeouts-main.md (hunk 1 above)
Comment (shellygr) [reply docs/user-guide/timeouts/timeouts-main.md:224]: didn't know about it!

File: docs/user-guide/timeouts/timeouts-main.md (hunk 3 above)
Comment (shellygr) [reply docs/user-guide/timeouts/timeouts-main.md:229]: I think it's pretty important distinction that some function that was already summarized will not have its summary overridden

File: docs/user-guide/timeouts/timeouts-main.md (hunk 4 above)
Comment (shellygr) [reply docs/user-guide/timeouts/timeouts-main.md:236]: in any case it's probably not very useful for users to compute on their own

File: docs/prover/cli/options.md [hunk 7]
```diff
@@ -513,6 +513,47 @@ certoraRun Bank.sol --verify Bank:Bank.spec --summary_recursion_limit 3
 ```
 
 
+(--auto_nondet_difficult_internal_funcs)=
+### `--auto_nondet_difficult_internal_funcs`
+
+**What does it do?**
+When this option is set, the Prover will auto-summarize
+view or pure internal functions that return a value type and are
```
Comment (alexandernutz) [inline docs/prover/cli/options.md:521]:
```suggestion
`view` or `pure` internal functions that return a value type and are
```
tioli :-)

File: docs/user-guide/timeouts/timeouts-main.md (hunk 3 above)
Comment (alexandernutz) [reply docs/user-guide/timeouts/timeouts-main.md]: right -- I like it better with the extra sentence that specifies it a bit more

**Review by alexandernutz — APPROVED** [review]
great!

**Pull request merged**